File: GhostBot/functions/fairy.py
from __future__ import annotations

import asyncio
from typing import TYPE_CHECKING

from GhostBot.config import FairyConfig
from GhostBot.functions.runner import Locational
from GhostBot.lib.math import linear_distance
from GhostBot.lib.talisman_ui_locations import TeamLocations

if TYPE_CHECKING:
    from GhostBot.controller.bot_controller import BotClientWindow


class Fairy(Locational):

    _team_members: dict[int, BotClientWindow] = {}

    def __init__(self, bot_controller, client: BotClientWindow):
        super().__init__(client)
        self.config: FairyConfig = client.config.fairy
        self._bot_controller = bot_controller

    async def _run(self) -> bool:
        await self._heal_self()
        for index, member in sorted(self._detect_team_members().items(), key=lambda i: i[1].hp_percent, reverse=True):
            if member.hp_percent < self.config.heal_team_threshold and linear_distance(self._client.location, member.location) < 20:
                await self._heal_team_member(index, member)

        await self._goto_start_location()
        return True

    async def _heal_self(self):
        if self._client.hp_percent < self.config.heal_self_threshold:
            await self._client.left_click(TeamLocations[0])
            self._client.press_key(self.config.bindings.get('heal'))
            self._log_debug(f'heal self')

    async def _heal_team_member(self, index: int, member: BotClientWindow):
        self._log_debug(f'Weak member {member.name} {member.hp_percent}')
        while member.hp_percent < 0.9 and self._client.running:
            await self._client.dismount()
            await self._client.close_inventory()
            await self._client.left_click(TeamLocations[index + 1])
            self._client.press_key(self.config.bindings.get('heal'))
            await asyncio.sleep(0.5)
        self._log_debug(f'{member.name}: healed')

    def _detect_team_members(self) -> dict[int, BotClientWindow]:
        """
        :return: a dict of {index: ExtendedClient} representing the current team members.
        """
        return {i: self._bot_controller.clients.get(name) for i, name in enumerate(self._client.team_members)}

File: kotlin-server/src/main/kotlin/com/ghostbot/functions/Fairy.kt
package com.ghostbot.functions

import com.ghostbot.controller.BotClientWindow
import com.ghostbot.controller.BotController
import com.ghostbot.lib.TeamLocations
import com.ghostbot.lib.linearDistance
import com.ghostbot.lib.sleepSeconds

/**
 * Port of GhostBot/functions/fairy.py.
 */
class Fairy(private val botController: BotController, client: BotClientWindow) :
    Locational(client) {

    private val config = client.config?.fairy ?: error("fairy config required")

    override fun _run(): Boolean {
        val selfThreshold = (config.healSelfThreshold as? Double) ?: 0.5
        val teamThreshold = (config.healTeamThreshold as? Double) ?: 0.5
        if (client.hpPercent() < selfThreshold) {
            healSelf(selfThreshold)
        }
        val members = detectTeamMembers().toList()
            .sortedByDescending { it.second.hpPercent() }
            .associate { it.first to it.second }
        for ((index, member) in members) {
            if (member.hpPercent() < teamThreshold &&
                linearDistance(client.location.x to client.location.y, member.location.x to member.location.y) < 20
            ) {
                healTeamMember(index, member)
            }
        }

        gotoStartLocation()
        return true
    }

    private fun healSelf(selfThreshold: Double) {
        while (client.hpPercent() < 0.9) {
            if (client.hpPercent() < selfThreshold) {
                _logInfo("Healing self...")
                client.leftClick(TeamLocations[0])
                client.pressKey((config.bindings as? Map<*, *>)?.get("heal"))
            }
        }
    }

    private fun healTeamMember(index: Int, member: BotClientWindow) {
        _logInfo("Healing Weak member ${member.name}")
        while (member.hpPercent() < 0.9 && client.running) {
            client.dismount()
            client.closeInventory()
            client.leftClick(TeamLocations[index + 1])
            client.pressKey((config.bindings as? Map<*, *>)?.get("heal"))
            sleepSeconds(0.5)
        }
        _logDebug("${member.name}: healed")
    }

    /**
     * @return a map of {index: client} representing the current team members.
     */
    private fun detectTeamMembers(): Map<Int, BotClientWindow> {
        val result = mutableMapOf<Int, BotClientWindow>()
        for ((i, name) in client.teamMembers.withIndex()) {
            val member = botController.clients[name] ?: continue
            if (!member.disconnected) result[i] = member
        }
        return result
    }
}
